chainercv/datasets/siamese_dataset.py

```diff
@@ -0,0 +1,119 @@
+import collections
+import numpy as np
+
+import chainer
+
+
+def _construct_label_to_key(labels):
+    d = collections.defaultdict(list)
+    for i, label in enumerate(labels):
+        d[label].append(i)
+    return d
+
+
+class SiameseDataset(chainer.dataset.DatasetMixin):
+
+    """A dataset that returns samples fetched from two datasets.
+
+    The dataset returns samples from the two base datasets.
+    If :obj:`pos_ratio` is not :obj:`None`,
+    :class:`SiameseDataset` can be configured to return positive
+    pairs at the ratio of :obj:`pos_ratio` and negative pairs at the ratio
+    of :obj:`1 - pos_ratio`.
+    In this mode, the base datasets are assumed to be label datasets that
+    return an image and a label as a sample.
+
+    Args:
+        dataset_0: The first base dataset.
+        dataset_1: The second base dataset.
+        pos_ratio (None or float): If this is not :obj:`None`,
+            this dataset tries to construct positive pairs at the
+            given rate. If :obj:`None`,
+            this dataset randomly samples examples from the base
+            datasets. The default value is :obj:`None`.
+        length (None): The length of this dataset. If :obj:`None`,
+            the length of the first base dataset is the length of this
+            dataset.
+        labels_0 (None or numpy.ndarray): The labels associated to
+            the first base dataset. The length should be the same as
+            the length of the first dataset. If this is :obj:`None`,
+            the labels are automatically fetched using the following
+            line of code: :obj:`[ex[1] for ex in dataset_0]`.
+            Also, if :obj:`pos_ratio` is :obj:`None`, this value
+            is ignored. The default value is :obj:`None`.
```

Comment:
How about describing the intention of this argument?
```
...
the labels are automatically fetched using the following line of code: :obj:`[ex[1] for ex in dataset_0]`.
If an array of labels already exists, users can skip this fetching iteration and reduce computation cost.
...
```

Reply:
Even if `labels_0` and `labels_1` are specified, `_construct_label_to_key` iterates these arrays. So, I think the effect of this argument is limited. Is this argument necessary?

Reply:
`_construct_label_to_key` would not iterate over the dataset, so no images would be loaded if `labels_*` are specified.

Reply:
I see.